File: lib/ci/queue/static.rb
module CI
  module Queue
    class Static
      attr_reader :progress, :total

      def initialize(tests, max_requeues: 0, requeue_tolerance: 0.0)
        @queue = tests
        @progress = 0
        @total = tests.size
        @max_requeues = max_requeues
        @global_max_requeues = (tests.size * requeue_tolerance).ceil
      end

      def to_a
        @queue.dup
      end

      def size
        @queue.size
      end

      def poll
        while test = @queue.shift
          yield test
          @progress += 1
        end
      end

      def empty?
        @queue.empty?
      end

      def acknowledge(test)
        true
      end

      def requeue(test)
        return false unless should_requeue?(test)
        requeues[test] += 1
        @queue.unshift(test)
        true
      end

      private

      attr_reader :max_requeues, :global_max_requeues

      def should_requeue?(test)
        requeues[test] < max_requeues && requeues.values.inject(0, :+) < global_max_requeues
      end

      def requeues
        @requeues ||= Hash.new(0)
      end
    end
  end
end

File: python/ciqueue/static.py
import math


class Static(object):
    distributed = False

    def __init__(self, tests, max_requeues=0, requeue_tolerance=0):
        self.queue = tests
        self.progress = 0
        self.total = len(tests)
        self.max_requeues = max_requeues
        self.global_max_requeues = math.ceil(requeue_tolerance * self.total)
        self.requeues = {}

    def __len__(self):
        return len(self.queue)

    def __iter__(self):
        while self.queue:
            yield self.queue.pop(0)
            self.progress += 1

    def acknowledge(self, test):  # pylint: disable=no-self-use,unused-argument
        return True

    def requeue(self, test):
        if self.requeues.get(test, 0) >= self.max_requeues:
            return False
        if sum(self.requeues.values()) >= self.global_max_requeues:
            return False

        self.requeues[test] = self.requeues.get(test, 0) + 1
        self.queue.insert(0, test)
        return True
